### PR TITLE
ENH: introduce spans missing trace group counter

### PR DESCRIPTION
**Pull request by chenqi0805**
*Issue #, if available:*

*Description of changes:*

-----

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.

### CODE REVIEW

File: data-prepper-plugins/otel-trace-group-prepper/src/test/java/com/amazon/dataprepper/plugins/prepper/oteltracegroup/OTelTraceGroupPrepperTests.java [hunk 1]
```diff
@@ -127,6 +134,9 @@ public void testTraceGroupFillSuccess() throws IOException {
         // Arrange
         Record<String> testRecord = buildRawSpanRecord(TEST_RAW_SPAN_MISSING_TRACE_GROUP_JSON_FILE_1);
         List<Record<String>> testRecords = Collections.singletonList(testRecord);
+        final List<Measurement> spansMissingTraceGroupMeasures = MetricsTestUtil.getMeasurementList(
+                new StringJoiner(MetricNames.DELIMITER).add(TEST_PIPELINE_NAME).add(PLUGIN_NAME)
+                        .add(OTelTraceGroupPrepper.SPANS_MISSING_TRACE_GROUP_INFO).toString());
```
Comment (wrijeff) [inline data-prepper-plugins/otel-trace-group-prepper/src/test/java/com/amazon/dataprepper/plugins/prepper/oteltracegroup/OTelTraceGroupPrepperTests.java:139]: nit - I see 40+ usages of `getMeasurementList()` all requiring users to create a StringJoiner and know how to assemble the string. Might be time to create utility method which takes in pipeline/plugin/metric Strings, does the joining, and returns the measurement object(s).

This highlighted block could also be split out into a private method as it's repeated 4 times in this class.

File: data-prepper-plugins/otel-trace-group-prepper/src/main/java/com/amazon/dataprepper/plugins/prepper/oteltracegroup/OTelTraceGroupPrepper.java [hunk 2]
```diff
@@ -88,6 +93,7 @@ public OTelTraceGroupPrepper(final PluginSetting pluginSetting) {
                 }
             } else {
                 recordsOut.add(record);
+                spanMissingTraceGroupCounter.increment();
                 final String spanId = (String) rawSpanMap.get(OTelTraceGroupPrepperConfig.SPAN_ID_FIELD);
                 LOG.info("Failed to find traceGroup for spanId: {} due to traceGroup missing for traceId: {}", spanId, traceId);
```
Comment (wrijeff) [inline data-prepper-plugins/otel-trace-group-prepper/src/main/java/com/amazon/dataprepper/plugins/prepper/oteltracegroup/OTelTraceGroupPrepper.java:98]: Not part of this PR, however I just noticed we might want this to be a WARN instead of INFO.

File: data-prepper-plugins/otel-trace-group-prepper/src/test/java/com/amazon/dataprepper/plugins/prepper/oteltracegroup/OTelTraceGroupPrepperTests.java (hunk 1 above)
Comment (chenqi0805) [reply data-prepper-plugins/otel-trace-group-prepper/src/test/java/com/amazon/dataprepper/plugins/prepper/oteltracegroup/OTelTraceGroupPrepperTests.java:139]: I agree. Let me work on the second part (highlighted block splitting) and reserve the first in a separate PR for cleaner workflow path.

File: data-prepper-plugins/otel-trace-group-prepper/src/main/java/com/amazon/dataprepper/plugins/prepper/oteltracegroup/OTelTraceGroupPrepper.java [hunk 3]
```diff
@@ -36,18 +36,23 @@
 @DataPrepperPlugin(name = "otel_trace_group_prepper", type = PluginType.PREPPER)
 public class OTelTraceGroupPrepper extends AbstractPrepper<Record<String>, Record<String>> {
 
+    public static final String SPANS_MISSING_TRACE_GROUP_INFO = "spansMissingTraceGroupInfo";
```
Comment (wrijeff) [inline data-prepper-plugins/otel-trace-group-prepper/src/main/java/com/amazon/dataprepper/plugins/prepper/oteltracegroup/OTelTraceGroupPrepper.java:39]: I should've noted this earlier, but do we want more metrics than just failures? Some that come to mind are:
1. Records in which are missing trace group
2. Records out which were successfully fixed
3. Records out which were unsuccessfully fixed (which is what the highlighted line is)

Would also suggest renaming SPANS_MISSING_TRACE_GROUP_INFO to something more indicative that a failure occurred - I could see spansMissingTraceGroupInfo being interpreted as either scenarios 1 or 3 above.

**Review by wrijeff — APPROVED** [review]
👍 

Left a comment/question about other metrics, doesn't have to block this PR.

File: data-prepper-plugins/otel-trace-group-prepper/src/main/java/com/amazon/dataprepper/plugins/prepper/oteltracegroup/OTelTraceGroupPrepper.java (hunk 3 above)
Comment (chenqi0805) [reply data-prepper-plugins/otel-trace-group-prepper/src/main/java/com/amazon/dataprepper/plugins/prepper/oteltracegroup/OTelTraceGroupPrepper.java:39]: Finer resolution sounds reasonable. Renaming should be necessary for that.

Comment (chenqi0805) [reply data-prepper-plugins/otel-trace-group-prepper/src/main/java/com/amazon/dataprepper/plugins/prepper/oteltracegroup/OTelTraceGroupPrepper.java]: @wrijeff Updated the metrics and also the README section. Might better receive another look.

**Review by wrijeff — APPROVED** [review]
LGTM 👍

**Review by kowshikn — APPROVED** [review]
Looks good to me.
